### PR TITLE
[dLLM] Reuse block KV/req slots in place across FDFO rounds

### PR DESCRIPTION
**Pull request by Hayden727**
## Motivation

Split out from #27551 per @ClawSeven's review suggestion to keep the FDFO framework PR focused and review the HBM-reuse optimization on its own.

In FDFO scheduling, an **unresolved** (continuing) block is re-denoised across rounds. Previously each round freed the block's KV slots and the `req_pool_idx`, then reallocated them the next round. This PR reuses that HBM in place instead.

### In-place KV / req-slot reuse
For an unresolved FDFO request, retain its `req_pool_idx` and the block's KV slots across FDFO rounds:
- `get_next_batch_to_run` only frees the req slot when the block actually resolves.
- `process_batch_result_dllm` no longer frees the unresolved block's KV.
- `alloc_for_extend` detects such requests (retained `req_pool_idx` + stashed `dllm_incomplete_ids`) and reuses the slots already mapped in `req_to_token`, allocating fresh slots only for the rest. Non-dLLM batches take the original allocate-everything path.

It's paging-safe: the dLLM scheduler forces `page_size` to a multiple of `block_size`, so each block is page-aligned and was validly paged when first created.

## Tests

Adds a CPU unit test (`test/registered/unit/mem_cache/test_dllm_fdfo_kv_reuse.py`) asserting the reuse lifecycle directly: an unresolved block reuses its mapped `req_to_token` slots (allocator queried only for fresh requests), and an all-reuse batch allocates nothing.

## Correctness (H200, flashinfer, page_size=32)

- **unit**: `test/registered/unit/mem_cache/test_dllm_fdfo_kv_reuse.py` — 2 passed.
- **bs=1 parity**: FDFO output is byte-identical to `--no-dllm-fdfo` (sync) on `test_dllm_fdfo_joint_threshold` (LLaDA2.1-mini).
- **bs=4 gsm8k, LLaDA2.0-mini, 200 examples**: FDFO 0.915 vs sync 0.930 — within run-to-run noise, no material accuracy regression.
- Full 200-example GSM8K runs completed without OOM.

## Performance — controlled A/B vs current base (#27551 already merged)

Same exclusive H200 GPU, **A/B interleaved per run**, one warmup eval discarded per launch, 3 measured runs, median output throughput. GSM8K, 200 examples, LowConfidence, `--dllm-fdfo`, `page_size=32`.

Base: `271e5ef5c3ac5733908ce270d556985c1dee0624`  
This PR: `0102ffee689781bc1caa7b97383ee55ad3832030`

| batch size | base (tok/s) | this PR (tok/s) | speedup |
|---|---:|---:|---:|
| 4  | 696.7 | 719.0 | **+3.2%** |
| 16 | 1468.2 | 1477.9 | +0.7% (within noise) |

The measurable gain is at small batch size, where avoiding unresolved-block KV/req-slot free/realloc churn is a larger fraction of total work. At bs=16 the delta is within run-to-run noise.

## Note

This PR now contains only the FDFO HBM-reuse optimization on top of main. The earlier in-place re-loop experiment was removed after profiling showed it increased actual forward calls and hurt throughput.

---
### CI States

Latest PR Test (Base): :x: [Run #29692317914](https://github.com/sgl-project/sglang/actions/runs/29692317914)
Latest PR Test (Extra): :x: [Run #29692317791](https://github.com/sgl-project/sglang/actions/runs/29692317791)

### CODE REVIEW

File: test/registered/unit/mem_cache/test_dllm_fdfo_kv_reuse.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+import unittest
+from types import SimpleNamespace
+
+import torch
+
+from sglang.srt.mem_cache.common import _alloc_extend_loc_with_kv_reuse
+from sglang.srt.mem_cache.memory_pool import ReqToTokenPool
+from sglang.test.ci.ci_register import register_cpu_ci
+
+register_cpu_ci(est_time=5)
```
Comment (BBuf) [inline test/registered/unit/mem_cache/test_dllm_fdfo_kv_reuse.py]: Could you specify the CPU CI suite here? The current registry parser requires exactly one of `suite` or `(stage, runner_config)`, so this currently fails `check-changes` before any tests run. For this unit test, something like `register_cpu_ci(est_time=5, suite="base-a-test-cpu")` should work.

File: python/sglang/srt/managers/scheduler.py [hunk 2]
```diff
@@ -2638,7 +2638,8 @@ def get_next_batch_to_run(
                 if self.dllm_config.first_done_first_out_mode:
                     if not req.dllm_incomplete_ids:
                         self.stash_chunked_request(req)
-                    self.req_to_token_pool.free(req)
+                        self.req_to_token_pool.free(req)
+                    # Otherwise, keep req slot/KV for reuse.
```
Comment (BBuf) [inline python/sglang/srt/managers/scheduler.py:2692]: Now that unresolved FDFO requests keep both `req_pool_idx` and KV alive across rounds, can we also update the abort path to explicitly clean up requests owned by `dllm_manager.waiting_queue` / `dllm_manager.staging_queue`? `abort_request()` currently scans the main waiting queue, running/last batches, and `chunked_req`, but not the dLLM manager queues. If an unresolved FDFO request is aborted while it is retained there, the req slot/KV can stay allocated until the request naturally resolves.

File: python/sglang/srt/mem_cache/common.py [hunk 3]
```diff
@@ -524,6 +544,79 @@ def alloc_for_extend(
...
+    reuse_dtype = fresh_slots.dtype if fresh_slots is not None else torch.int64
+    parts: list[torch.Tensor] = []
+    fresh_ptr = 0
+    for i in range(len(reuse_kv)):
+        prefix_len = int(prefix_lens_cpu[i])
+        extend_len = int(extend_lens_cpu[i])
+        if reuse_kv[i]:
+            req_idx = int(req_pool_indices_cpu[i])
+            # Copy because write_cache_indices mutates this row.
+            parts.append(
```
Comment (BBuf) [inline python/sglang/srt/mem_cache/common.py]: Can we guard against partial reuse here? `add_dllm_staging_req()` can truncate the next staging extend length based on the remaining token budget, but the retained unresolved block was originally a full dLLM block. If `extend_len < block_size`, this only reuses the prefix of the retained KV range and the tail becomes unreachable without being freed. I think we should either assert/guarantee full-block reuse for retained FDFO requests, or explicitly free the unused retained tail.

File: test/registered/unit/mem_cache/test_dllm_fdfo_kv_reuse.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+        allocator = _FakeAllocator(base=200)
+        batch = _make_batch(self.pool, allocator)
+
+        reuse_kv = [True, False]
+        req_pool_indices_cpu = torch.tensor([1, 2], dtype=torch.int64)
+        prefix_lens_cpu = torch.tensor([prefix_len, prefix_len], dtype=torch.int64)
+        extend_lens_cpu = torch.tensor([b, b], dtype=torch.int64)
+
+        out = _alloc_extend_loc_with_kv_reuse(
```
Comment (BBuf) [inline test/registered/unit/mem_cache/test_dllm_fdfo_kv_reuse.py]: This is a useful unit-level check, but it only exercises the private helper with `page_size=1`. Could we add at least one test through `alloc_for_extend()` so it covers `alloc_req_slots()` and `write_cache_indices()` as well? The risky cases here are mixed fresh+reuse, all-reuse, `page_size > 1`, and abort/truncated reuse lifecycle.

**Review by BBuf — CHANGES_REQUESTED** [review]
Thanks for splitting this optimization out. The in-place reuse direction makes sense to me, but I think we need to fix the CI registration issue and tighten the retained-KV lifecycle before merging, especially abort cleanup and partial/truncated reuse.

**Review by BBuf — APPROVED** [review]
